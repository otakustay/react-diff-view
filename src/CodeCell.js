import {PureComponent} from 'react';

export default class CodeCell extends PureComponent {

    static defaultProps = {
        onRender() {
        }
    };

    cell = null;

    componentDidMount() {
        this.notifyRender();
    }

    componentDidUpdate(prevProps) {
        if (prevProps.text !== this.props.text || prevProps.html !== this.props.html) {
            this.notifyRender();
        }
    }

    notifyRender() {
        this.props.onRender(this.cell);
    }

    render() {
        /* eslint-disable no-unused-vars */
        const {text, html, onRender, ...props} = this.props;
        /* eslint-enable no-unused-vars */
        const ref = cell => (this.cell = cell);

        if (html) {
            /* eslint-disable react/no-danger */
            return <td ref={ref} {...props} dangerouslySetInnerHTML={{__html: html}} />;
            /* eslint-enable react/no-danger */
        }

        return <td ref={ref} {...props}>{text}</td>;
    }
}
