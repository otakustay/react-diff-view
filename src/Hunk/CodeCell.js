import {memo} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const defaultRenderToken = ({type, value, markType, properties, children}, i) => {
    const renderWithClassName = className => (
        <span key={i} className={className}>
            {value ? value : (children && children.map(defaultRenderToken))}
        </span>
    );

    switch (type) {
        case 'text':
            return value;
        case 'mark':
            return renderWithClassName(`diff-code-mark diff-code-mark-${markType}`);
        case 'edit':
            return renderWithClassName('diff-code-edit');
        default:
            return renderWithClassName(classNames(properties.className));
    }
};

const CodeCell = props => {
    const {text, tokens, renderToken, ...attributes} = props;
    const actualRenderToken = renderToken
        ? (token, i) => renderToken(token, defaultRenderToken, i)
        : defaultRenderToken;

    return (
        <td {...attributes}>
            {
                tokens
                    ? (tokens.length ? tokens.map(actualRenderToken) : ' ')
                    : (text || ' ')
            }
        </td>
    );
};

CodeCell.propTypes = {
    text: PropTypes.string.isRequired,
    tokens: PropTypes.arrayOf(PropTypes.object),
};

CodeCell.defaultProps = {
    tokens: null,
};

export default memo(CodeCell);
