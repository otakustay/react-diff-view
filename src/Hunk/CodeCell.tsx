// @ts-nocheck ignore: third-party
import {memo} from 'react';
import classNames from 'classnames';

const defaultRenderToken = ({type, value, markType, properties, className, children}, i) => {
    const renderWithClassName = className => (
        <span key={i} className={className}>
            {value ? value : (children && children.map(defaultRenderToken))}
        </span>
    );


    switch (type) {
        case 'text':
            return value;
        case 'mark':
            return renderWithClassName(`diff-code-mark diff-code-mark-${markType}`);
        case 'edit':
            return renderWithClassName('diff-code-edit');
        default: {
            // properties normally not exist since it is deconstructed in pickRange, remove in next major release
            const legacyClassName = properties && properties.className;
            return renderWithClassName(classNames(className || legacyClassName));
        }
    }
};

function CodeCell(props) {
    const {text, tokens, renderToken, ...attributes} = props;
    const actualRenderToken = renderToken
        ? (token, i) => renderToken(token, defaultRenderToken, i)
        : defaultRenderToken;

    return (
        <td {...attributes}>
            {
                tokens
                    ? (tokens.length ? tokens.map(actualRenderToken) : ' ')
                    : (text || ' ')
            }
        </td>
    );
}

export default memo(CodeCell);
