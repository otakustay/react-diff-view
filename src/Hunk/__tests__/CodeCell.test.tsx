// @ts-nocheck ignore: third-party
import * as renderer from 'react-test-renderer';
import CodeCell from '../CodeCell';

describe('renderToken', () => {
    test('renders correctly', () => {
        expect(renderer.create(
            <CodeCell text="A" />
        ).toJSON()).toMatchSnapshot();
    });

    test('use renderToken', () => {
        const renderToken = (token, defaultRender, i) => {
            if (token.type === 'searchResult') {
                return (
                    <span key={i} className="search-result">
                        {token.children && token.children.map((token, i) => renderToken(token, defaultRender, i))}
                    </span>
                );
            }
            return defaultRender(token, i);
        };
        expect(renderer.create(
            // eslint-disable-next-line react/jsx-no-bind
            <CodeCell text="A" renderToken={renderToken} />
        ).toJSON()).toMatchSnapshot();
    });

    test('ref token', () => {
        const renderToken = (token, defaultRender, i) => {
            if (token.type === 'ref') {
                return (
                    <span key={i} className="ref-container">
                        {defaultRender(token, i)}
                    </span>
                );
            }
            return defaultRender(token, i);
        };
        const tokens = [{type: 'ref', className: 'ref'}];
        expect(renderer.create(
            // eslint-disable-next-line react/jsx-no-bind
            <CodeCell text="A" tokens={tokens} renderToken={renderToken} />
        ).toJSON()).toMatchSnapshot();
    });
});
