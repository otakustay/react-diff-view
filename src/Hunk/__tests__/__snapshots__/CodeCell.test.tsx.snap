// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderToken ref token 1`] = `
<td>
  <span
    className="ref-container"
  >
    <span
      className="ref"
    />
  </span>
</td>
`;

exports[`renderToken renders correctly 1`] = `
<td>
  A
</td>
`;

exports[`renderToken use renderToken 1`] = `
<td>
  A
</td>
`;
